src/lightning/fabric/fabric.py

```diff
@@ -596,16 +599,27 @@ def init(self) -> Generator:
         The parameters get created on the device (if using PyTorch 2.0 or newer) and with the right data type right away
         without wasting memory being allocated unnecessarily.
         """
-        with self._strategy.init_context():
```

Comment:
Because the strategy in `module_init_context()` does more than just set the dtype and device. For example, we don't want that the deepspeed zero Init() context manager is being used when we do

```py
with fabric.init():
    torch.ones(2, 2)
```

or other side effects from the strategy. The two IMO features should be completely disentangled from each other.

Reply:
Let me phrase it in another way. I want that
1. init_tensor() is for tensors only (device + dtype). For this, we don't need an extra hook on the strategy because it's just a simple with `self.device` call that generalizes for all tensors.
2. init_module() is for module instantiation only. should do whatever the strategy wants. This deserves its own hook where the strategy can put their own context managing logic. For the non-distributed strategies, the code is trivial and there is no difference to what init_tensor() does. But for deepspeed and fsdp, we need to have custom logic. 

I want to disentangle the two completely (I insist!) so that I can implement empty weight initialization in #17616 in `init_module()` only.